File: postorder-inorder.py
# Time Complexity: O(n)
# Space Complexity: O(n)
# Approach: 
# Get the value of the root node from the postorder array (last element - left, right, root)
# Find the index of root element in inorder array
# Recursively construct the left and right subtrees using the inorder array

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        # Hashmap to access index of root from inorder array in O(1)
        ino_map = {}
        for i in range(len(inorder)):
            ino_map[inorder[i]] = i
            
        
        def helper(p_start, p_end, i_start, i_end):
            # Base Case
            if p_start > p_end:
                return None
            if p_start == p_end:
                return TreeNode(postorder[p_start])
            
            # Root is the last element in postorder
            root_val = postorder[p_end]
            root_index = ino_map[root_val]
            
            # create a new treenode
            root = TreeNode(root_val)
            
            # Calculate number of elements in the left and right subtrees
            numleft = root_index - i_start
            numright = i_end - root_index
            
            # recursively construct the left and right subtrees
            root.left = helper(p_start,
                               p_start + numleft - 1,
                               i_start,
                               root_index - 1)
            root.right = helper(p_start + numleft,
                               p_end - 1,
                               root_index + 1,
                               i_end)
            return root
        
        return helper(0, len(postorder) - 1, 0, len(inorder) - 1)

File: PostOrder_Inorder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //Time O(N)
    //Space O(N)
    HashMap<Integer, Integer> map;
    int index;
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(postorder.length == 0)
        {
            return null;
        }
        map = new HashMap<>();
        index = postorder.length-1;
        for(int i=0 ; i<inorder.length ; i++)
        {
            map.put(inorder[i] , i);
        }
        return helper(postorder, 0 , postorder.length-1);
    }
    public TreeNode helper(int[] postorder, int start, int end)
    {
        if(start > end || index < 0)
        {
            return null;
        }
        int inIndex = map.get(postorder[index]);
        TreeNode root = new TreeNode(postorder[index]);
        index--;
        root.right = helper(postorder, inIndex+1, end);
        //In post order you will visit the root of the right subtree unlike the left subtree of preorder
        root.left = helper(postorder , start, inIndex-1);
        
        return root;
    }
}